pantab/_writer.py

```diff
@@ -30,8 +30,48 @@ def _timedelta_to_interval(td: pd.Timedelta) -> tab_api.Interval:
 
 
 def _validate_table_mode(table_mode: str) -> None:
-    if table_mode not in {"a", "w"}:
-        raise ValueError("'table_mode' must be either 'w' or 'a'")
+    if table_mode not in {"a", "w", "i"}:
+        raise ValueError("'table_mode' must be either 'w', 'a' or 'i'")
+
+
+def _validate_key_exists(table_key: str, columns: pd.Index) -> None:
+    if table_key is None:
+        raise ValueError("'table_key' must be provided when using table_mode='i'.")
+    if table_key not in columns:
+        raise LookupError(f"Key '{table_key}' does not exist in the table.")
+
+
+def _compare_frames_keys_len(frames_len: int, keys_len: int) -> None:
+    if not (frames_len == keys_len):
+        raise ValueError("Number of frames does not match number of keys.")
+
+
+def _filter_new_records(
+    df: pd.DataFrame,
+    connection: tab_api.Connection,
+    table: pantab_types.TableType,
+    table_key: str,
+) -> Callable:
+    """
+    Compares unique values from an incoming DataFrame based on a given key, with
+    the values in an existing tableau Hyper table and returns a filtered DataFrame
+    excluding those already contained within the table.
+    """
+    _validate_key_exists(table_key, df.columns)
+
+    if isinstance(table, str):
+        table = tab_api.TableName(table)
+
+    with connection.execute_query(
+        query=f"""
+               select distinct {tab_api.escape_name(table_key)}
+               from {table}
+               """
+    ) as keys:
+        existing_keys = set(itertools.chain(*list(keys)))
```

Comment:
It might be more performant to create a DataFrame from the result, use `pd.merge(..., indicator=True)` and then just filter on the appropriate indicator